python/tank/context.py

```diff
@@ -624,37 +623,105 @@ def _fields_from_template_tree(self, template, fields, entities):
         path_cache = PathCache(self.__tk)
 
         # Step 3 - walk templates from the root down,
-        # for each template, get all paths we have stored in the database
-        # and find any fields we can for it
+        # for each template, get all paths we have stored in the database and find any fields we can for it, making 
+        # sure that none of the found fields conflict with the list of entities provided to this method
         try:
             # build up a list of fields as we go so that each level matches
-            # at least the fields from the previous level
+            # at least the fields from all previous levels
             found_fields = {}
 
-            for cur_template in templates:
-                for key in cur_template.keys.values():
+            for template in templates:
+                for key_name in template.keys.keys():
```

Comment:
Minor: The fact that we use the term template "key" (and therefore `template.keys`) makes the code (`keys.keys()`) a bit confusing! I suggest adding a comment pointing out the keys data structure just to help

```
# template.keys is a dictionary of template key objects keyed by name
for key_name in template.keys.keys():
```